### PR TITLE
[SPARK-47415][SQL] Collation support: Levenshtein

### PR DESCRIPTION
**Pull request by nikolamand-db**
### What changes were proposed in this pull request?

Levenshtein distance requires modifications to accommodate collated inputs properly:
- Allow collated inputs in `inputTypes`
- Require same collations for input types in `CollationTypeCasts`
- Comparison of byte substrings of `UTF8String` for equality is not enough anymore; we split checks of substrings (character) equality into classes for binary equality and semantic equality, latter is used with collations which don't support binary equality checks
- Each collation gets substring equality checks statically in collation factory to reduce object allocation on each Levenshtein distance evaluation

### Why are the changes needed?

To support Levenshtein distance expression with collations.

### Does this PR introduce _any_ user-facing change?

Yes, it changes behavior of Levenshtein distance expression when evaluated on collated strings.

### How was this patch tested?

Added checks to `CollationStringExpressionsSuite` in addition to handful of existing Levenshtein checks.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by nikolamand-db** [comment]
Please review collation team @dbatomic @stefankandic @uros-db @mihailom-db @stevomitric.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -2213,8 +2213,8 @@ case class Levenshtein(
   }
 
   override def inputTypes: Seq[AbstractDataType] = threshold match {
-    case Some(_) => Seq(StringType, StringType, IntegerType)
-    case _ => Seq(StringType, StringType)
+    case Some(_) => Seq(StringTypeAnyCollation, StringTypeAnyCollation, IntegerType)
+    case _ => Seq(StringTypeAnyCollation, StringTypeAnyCollation)
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2217]: Maybe we should special case this in CollationTypeCasts. We do not want to have an expression on the place of third parameter with StringType and take it into account, as it will explicitly be casted to Integer, and technically it's collation should not be counted. Something similar to If expression that is special cased.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 2]
```diff
@@ -89,6 +89,73 @@ class CollationStringExpressionsSuite
...
+      CollationTestCase("Xü", "Ü", "UTF8_BINARY_LCASE", 1),
+      CollationTestCase("Xũ", "Üx", "UTF8_BINARY_LCASE", 2),
+      CollationTestCase("", "something", "UTF8_BINARY_LCASE", 9),
+      CollationTestCase("sOmeThINg", "SOMETHING", "UTF8_BINARY_LCASE", 0),
+      CollationTestCase("sOmeThINg", "SOMETHING", "UNICODE", 5),
+      CollationTestCase("sOmeThINg", "SOMETHING", "UNICODE_CI", 0)
+      // scalastyle:on nonascii
+    ).foreach(c => checkEvaluation(prepareLevenshtein(c.s1, c.s2, c.collation), c.expectedResult))
+  }
+
```
Comment (mihailom-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:121]: @cloud-fan Should we also add a test to CollationSuite Implicit casting to check if casting actually works for this expression?

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:121]: my thoughts are: yes

@dbatomic and myself will be finishing some refactoring PRs soon, and will likely request all PRs to comply with some new design before further reivew

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 3]
```diff
@@ -1509,12 +1509,66 @@ public boolean semanticEquals(final UTF8String other, int collationId) {
...
+    CollationSubstringEquals(int collationId) {
+      this.collationId = collationId;
+      this.left = new UTF8String();
+      this.right = new UTF8String();
+    }
+
+    @Override
+    public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight) {
+      this.left.base = left.base;
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1544]: can we maybe extract this logic into a method?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 4]
```diff
@@ -89,6 +89,73 @@ class CollationStringExpressionsSuite
...
+        right: String,
+        collation: String): Levenshtein = {
+      val collationId = CollationFactory.collationNameToId(collation)
+      val leftLit = Literal.create(left, StringType(collationId))
+      val rightLit = Literal.create(right, StringType(collationId))
+      Levenshtein(leftLit, rightLit)
+    }
+
+    Seq(
+      // scalastyle:off nonascii
```
Comment (stefankandic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:104]: there's a good chance we will use nonascii characters a lot in this file, what about disabling this for the entire class/file?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 5]
```diff
@@ -1509,12 +1509,66 @@ public boolean semanticEquals(final UTF8String other, int collationId) {
...
+        return false;
+      }
+      return (ByteArrayMethods.arrayEquals(left.base, left.offset + posLeft, right.base,
+        right.offset + posRight, lenLeft));
+    }
+  }
+
+  private static final ByteSubstringEquals byteSubstringEquals = new ByteSubstringEquals();
+
+  private static class CollationSubstringEquals implements SubstringEquals {
```
Comment (stefankandic) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1531]: have you actually tested the impact of having a smaller number of allocations? Because creating a UTF8 string might not be that expensive, especially compared to converting from UTF8String to String which will happen anyway

Comment (nikolamand-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1531]: It's true that `UTF8String` is a small object (base, offset and length), but the issue with Levenshtein distance is that the algorithm has quadratic complexity. So when comparing two strings of length 100, you would allocate 10k objects which is a waste.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 2 above)
Comment (nikolamand-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:120]: Added the tests to CollationSuite, please check.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 4 above)
Comment (nikolamand-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala]: Wrapped the class with `scalastyle:off nonascii`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 1 above)
Comment (nikolamand-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2217]: Added special case for Levenshtein to `CollationTypeCasts`, please check.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 6]
```diff
@@ -1509,12 +1515,62 @@ public boolean semanticEquals(final UTF8String other, int collationId) {
...
+    CollationSubstringEquals(int collationId) {
+      this.collationId = collationId;
+      this.left = new UTF8String();
+      this.right = new UTF8String();
+    }
+
+    @Override
+    public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+      int lenLeft, int lenRight) {
+      moveAddress(this.left, left.base, left.offset + posLeft, lenLeft);
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1550]: Should `moveAddress` be an instance method that takes two int parameters (posLeft and lenLeft)?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 7]
```diff
@@ -1636,14 +1699,13 @@ public int levenshteinDistance(UTF8String other, int threshold) {
         } else if (i == min - 1) {
           num_bytes_i = 0;
         } else {
-          if (ByteArrayMethods.arrayEquals(t.base, t.offset + j_bytes,
-                  s.base, s.offset + i_bytes, num_bytes_j)) {
+          num_bytes_i = numBytesForFirstByte(s.getByte(i_bytes));
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:1702]: I don't quite get the details here. Why do we need to pre-evalute `numBytesForFirstByte` now? It was evaluated at the end of the loop body before.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:120]: Yea we need both unit tests and end-to-end tests

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 8]
```diff
@@ -2210,7 +2210,7 @@ case class Levenshtein(
     threshold: Option[Expression] = None)
   extends Expression
   with ImplicitCastInputTypes
-  with NullIntolerant{
+  with NullIntolerant {
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2213]: nit: Remove unnecessary changes.

Comment (nebojsa-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Done

**Comment by nebojsa-db** [comment]
Please review latest iteration @dbatomic @stefankandic @uros-db @mihailom-db @stevomitric

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 9]
```diff
@@ -722,6 +722,65 @@ public static UTF8String execLowercase(
     }
   }
 
+  /**
+   * Utility class for collation aware Levenshtein function.
+   */
+  public static class Levenshtein{
```
Comment (nikolamand-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:728]:
```suggestion
  public static class Levenshtein {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 10]
```diff
@@ -722,6 +722,65 @@ public static UTF8String execLowercase(
...
+      }
+    }
+
+    public static Integer exec(final UTF8String left, final UTF8String right, final int collationId){
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+
+      if (collation.supportsBinaryEquality){
+        return left.levenshteinDistance(right);
+      }
+      else{
```
Comment (nikolamand-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:759]:
```suggestion
      else {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 11]
```diff
@@ -722,6 +722,65 @@ public static UTF8String execLowercase(
...
+        this.right.moveAddress(right, posRight, lenRight);
+        return CollationFactory.fetchCollation(collationId).equalsFunction
+                .apply(this.left, this.right);
+      }
+    }
+
+    public static Integer exec(final UTF8String left, final UTF8String right, final int collationId){
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+
+      if (collation.supportsBinaryEquality){
```
Comment (nikolamand-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:756]:
```suggestion
      if (collation.supportsBinaryEquality) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 12]
```diff
@@ -722,6 +722,65 @@ public static UTF8String execLowercase(
...
+      public boolean equals(UTF8String left, UTF8String right, int posLeft, int posRight,
+                            int lenLeft, int lenRight) {
+        this.left.moveAddress(left, posLeft, lenLeft);
+        this.right.moveAddress(right, posRight, lenRight);
+        return CollationFactory.fetchCollation(collationId).equalsFunction
+                .apply(this.left, this.right);
+      }
+    }
+
+    public static Integer exec(final UTF8String left, final UTF8String right, final int collationId){
```
Comment (nikolamand-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:753]:
```suggestion
    public static Integer exec(final UTF8String left, final UTF8String right, final int collationId) {
```

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 13]
```diff
@@ -722,6 +722,65 @@ public static UTF8String execLowercase(
...
+
+    public static Integer execWithThreshold(final UTF8String left, final UTF8String right, final int threshold, final int collationId){
+      CollationFactory.Collation collation = CollationFactory.fetchCollation(collationId);
+
+      if (collation.supportsBinaryEquality){
+        return left.levenshteinDistance(right, threshold);
+      }
+      else{
+        return left.levenshteinDistance(right, threshold, new CollationSubstringEquals(collationId));
+      }
```
Comment (nikolamand-db) [inline common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:772]: We need to split logic into `execBinary` and `execNonBinary` here, please look at examples such as https://github.com/apache/spark/pull/45963/files#diff-3095052ebd126ed810dceb295897d0d7b1fba11b19a4da371ab43a2f41e96313R465

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 14]
```diff
@@ -959,6 +959,54 @@ class CollationStringExpressionsSuite
...
+      LevenshteinThresholdTestCase("sOmeThINg", "SOMETHING", "UNICODE_CI", 10, 0)
+    ).foreach(c => {
+      val query = s"SELECT levenshtein(collate('${c.l}', '${c.c}'), " +
+        s"collate('${c.r}', '${c.c}'), ${c.t})"
+      // Result & data type
+      checkAnswer(sql(query), Row(c.result))
+      assert(sql(query).schema.fields.head.dataType.sameType(IntegerType))
+    })
+  }
+
```
Comment (nikolamand-db) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:1009]: We can have more relaxed testing in this file in general, these are fine. However, we should very thoroughly perform unit tests in https://github.com/apache/spark/blob/master/common/unsafe/src/test/java/org/apache/spark/unsafe/types/CollationSupportSuite.java for Levenshtein distance. Please look at those examples and add analogous tests.

**Review by nikolamand-db — COMMENTED** [review]
Minor concerns, otherwise looks great!

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 13 above)
Comment (nikolamand-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:772]: This should be done in other places as well (both threshold & non-threshold Levenshtein).

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 9 above)
Comment (nebojsa-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:728]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala [hunk 15]
```diff
@@ -71,6 +71,9 @@ object CollationTypeCasts extends TypeCoercionRule {
       val Seq(newStr, newPad) = collateToSingleType(Seq(str, pad))
       stringPadExpr.withNewChildren(Seq(newStr, len, newPad))
 
+    case lev: Levenshtein =>
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala:74]: nit: please use levExpr to follow other patterns

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM, except for nit changes

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 13 above)
Comment (nebojsa-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:772]: Done

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 6 above)
Comment (nebojsa-db) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: Done, in addition with baseString which is also needed.

**Comment by nikolamand-db** [comment]
After a lengthy discussion we decided to stop Levenshtein collations effort. The well defined definition of Levenshtein as _the minimum number of single-character edits to make the strings compare equal_ would be interpreted under collations as _the minimum number of single-character edits to make the strings equal under given collation_ is very hard to achieve under non-exponential time.

Reason for this is that Levenshtein distance algorithm assumes we can compare single character in both strings for equality whereas under collations we cannot safely assume it is enough to compare single character. For example, sequences `İ` and `i\u0307` compare equal under `UTF8_BINARY_LCASE` collation but have different number of characters (1 vs 2 in order).

Thanks @nebojsa-db and reviewers for your effort.
